### PR TITLE
ConnectionManager: reconnect on disconnect pushes.

### PR DESCRIPTION
**Pull request by nihohit**
This change allows the ConnectionManager to reconnect not only when receiving errors on requests, but also when receiving disconnect pushes on the push channel from the internal connection.

Based over https://github.com/redis-rs/redis-rs/pull/1392

### CODE REVIEW

**Comment by nihohit** [comment]
@avifenesh, another one.

**Comment by altanozlu** [comment]
Currently when there is an error on CM (Connection Manager), user will know about it but by receiving an error, with this change when push sender is given user might get an error or not, depending on how fast it triggers re-connect. That behaviour might is flaky and users might rely on wrong things.
If this is okay then I would like to ask, is it possible to handle it without using PushInfo::Disconnection so there will be no need for user setting up push sender for this feature and it can be the default behaviour.

**Comment by nihohit** [comment]
> when push sender is given user might get an error or not, depending on how fast it triggers re-connect. That behaviour might is flaky and users might rely on wrong things.

Can you give an example of a potentially problematic chain of events?

> If this is okay then I would like to ask, is it possible to handle it without using PushInfo::Disconnection so there will be no need for user setting up push sender for this feature and it can be the default behaviour.

You're asking whether the connection manager can automatically reconnect without setting a push manager?

**Comment by altanozlu** [comment]
> > when push sender is given user might get an error or not, depending on how fast it triggers re-connect. That behaviour might is flaky and users might rely on wrong things.
> 
> Can you give an example of a potentially problematic chain of events?
> 
Testing things in local and somehow it always reconnect and nothing is noticed
I can't come up with one now, but auto reconnect is best effort and if user sends a command when auto reconnect didn't happen yet, command will result in an error, maybe it should be documented somewhere ?

> > If this is okay then I would like to ask, is it possible to handle it without using PushInfo::Disconnection so there will be no need for user setting up push sender for this feature and it can be the default behaviour.
> 
> You're asking whether the connection manager can automatically reconnect without setting a push manager?

I wanted to know if it's feasible to have reconnections without user providing push sender.

**Comment by nihohit** [comment]
> I wanted to know if it's feasible to have reconnections without user providing push sender.

Well, I pushed a commit doing that, but it seems to fail for async_std runtime, when using RusTLS. not sure why.

**Comment by altanozlu** [comment]
> > I wanted to know if it's feasible to have reconnections without user providing push sender.
> 
> Well, I pushed a commit doing that, but it seems to fail for async_std runtime, when using RusTLS. not sure why.

It's something related to tls, for me tokio also hangs with:
```
RUSTFLAGS="-D warnings" REDISRS_SERVER_TYPE=tcp+tls RUST_BACKTRACE=1 cargo test basic_async::manager_should_reconnect_without_actions_if_push_sender_is_set::case_1_tokio --locked -p redis --all-feature
```
which is weird because test shouldn't even run something important because there is `if ctx.protocol == ProtocolVersion::RESP2 {` check

as far as I see `TestContext::new` hangs

**Comment by nihohit** [comment]
> as far as I see TestContext::new hangs

That would be weird, because this doesn't hang in any other test.

**Comment by nihohit** [comment]
in any case, this doesn't reproduce locally, but seems consistent in CI. annoying :(

**Comment by nihohit** [comment]
@altanozlu tests seem to pass now.

File: redis/src/aio/connection_manager.rs [hunk 1]
```diff
@@ -454,8 +511,9 @@ impl ConnectionManager {
     pub async fn subscribe(&mut self, channel_name: impl ToRedisArgs) -> RedisResult<()> {
         check_resp3!(self.client.connection_info.redis.protocol);
         let mut cmd = cmd("SUBSCRIBE");
-        cmd.arg(channel_name);
+        cmd.arg(&channel_name);
         cmd.exec_async(self).await?;
+
```
Comment (altanozlu) [inline redis/src/aio/connection_manager.rs:516]: nitpick: unnecessary new line

**Review by altanozlu — APPROVED** [review]
LGTM except one little newline which you can dismiss 🙌🏻

**Pull request merged**